source/s3_request.c

```diff
@@ -95,3 +124,259 @@ static void s_s3_request_destroy(void *user_data) {
+    AWS_PRECONDITION(metrics);
+    AWS_PRECONDITION(total_duration);
+    *total_duration = metrics->time_metrics.total_duration_ns;
+}
+
+int aws_s3_request_metrics_get_send_start_timestamp_ns(
+    const struct aws_s3_request_metrics *metrics,
+    uint64_t *send_start_time) {
+    AWS_PRECONDITION(metrics);
+    AWS_PRECONDITION(send_start_time);
+    if (metrics->time_metrics.send_start_timestamp_ns == 0) {
+        return aws_raise_error(AWS_ERROR_S3_METRIC_DATA_NOT_AVAILABLE);
+    }
+    *send_start_time = metrics->time_metrics.send_start_timestamp_ns;
+    return AWS_OP_SUCCESS;
+}
+
+int aws_s3_request_metrics_get_send_end_timestamp_ns(
+    const struct aws_s3_request_metrics *metrics,
+    uint64_t *send_end_time) {
+    AWS_PRECONDITION(metrics);
+    AWS_PRECONDITION(send_end_time);
+    if (metrics->time_metrics.send_end_timestamp_ns == 0) {
+        return aws_raise_error(AWS_ERROR_S3_METRIC_DATA_NOT_AVAILABLE);
+    }
+    *send_end_time = metrics->time_metrics.send_end_timestamp_ns;
+    return AWS_OP_SUCCESS;
+}
+
+int aws_s3_request_metrics_get_sending_duration_ns(
+    const struct aws_s3_request_metrics *metrics,
+    uint64_t *sending_duration) {
+    AWS_PRECONDITION(metrics);
+    AWS_PRECONDITION(sending_duration);
+    if (metrics->time_metrics.send_end_timestamp_ns == 0 || metrics->time_metrics.send_start_timestamp_ns == 0) {
+        return aws_raise_error(AWS_ERROR_S3_METRIC_DATA_NOT_AVAILABLE);
+    }
+    *sending_duration = metrics->time_metrics.sending_duration_ns;
+    return AWS_OP_SUCCESS;
+}
+
+int aws_s3_request_metrics_get_receive_start_timestamp_ns(
+    const struct aws_s3_request_metrics *metrics,
+    uint64_t *receive_start_time) {
+    AWS_PRECONDITION(metrics);
+    AWS_PRECONDITION(receive_start_time);
+    if (metrics->time_metrics.receive_start_timestamp_ns == 0) {
+        return aws_raise_error(AWS_ERROR_S3_METRIC_DATA_NOT_AVAILABLE);
+    }
+    *receive_start_time = metrics->time_metrics.receive_start_timestamp_ns;
+    return AWS_OP_SUCCESS;
+}
+
+int aws_s3_request_metrics_get_receive_end_timestamp_ns(
+    const struct aws_s3_request_metrics *metrics,
+    uint64_t *receive_end_time) {
+    AWS_PRECONDITION(metrics);
+    AWS_PRECONDITION(receive_end_time);
+    if (metrics->time_metrics.receive_end_timestamp_ns == 0) {
+        return aws_raise_error(AWS_ERROR_S3_METRIC_DATA_NOT_AVAILABLE);
+    }
+    *receive_end_time = metrics->time_metrics.receive_end_timestamp_ns;
+    return AWS_OP_SUCCESS;
+}
+
+int aws_s3_request_metrics_get_receiving_duration_ns(
+    const struct aws_s3_request_metrics *metrics,
+    uint64_t *receiving_duration) {
+    AWS_PRECONDITION(metrics);
+    AWS_PRECONDITION(receiving_duration);
+    if (metrics->time_metrics.receive_end_timestamp_ns == 0 || metrics->time_metrics.receive_start_timestamp_ns == 0) {
+        return aws_raise_error(AWS_ERROR_S3_METRIC_DATA_NOT_AVAILABLE);
+    }
+    *receiving_duration = metrics->time_metrics.receiving_duration_ns;
```

Comment:
wouldn't it be less error_prone to continue storing times as int64_t internally, so we could simply use -1 to indicate date not available?
```suggestion
    if (metrics->time_metrics.receiving_duration_ns < 0) {
        return aws_raise_error(AWS_ERROR_S3_METRIC_DATA_NOT_AVAILABLE);
    }
    *receiving_duration = (uint64_t)metrics->time_metrics.receiving_duration_ns;
```

vs doing this thing where 0 maaay be valid from some values, so we better consult two other values

like we were doing in that "Range: 0-0" bug we just had